File: Openwhisk/Javascript/rsa.js
/**
 * Simple RSA decryption algorithm
 */
function main(args) {

    if (!args.n) {
        return {
            statusCode: 500,
            headers: { 'Content-Type': 'application/json' },
            body: {
                "message": "The parameter 'n' must be defined!"
            }
        };
    }

    if (!args.e) {
        return {
            statusCode: 500,
            headers: { 'Content-Type': 'application/json' },
            body: {
                "message": "The parameter 'e' must be defined!"
            }
        };
    }

    if (!args.cipher) {
        return {
            statusCode: 500,
            headers: { 'Content-Type': 'application/json' },
            body: {
                "message": "The parameter 'cipher' must be defined!"
            }
        };
    }

    // arguments n, e, cipher
    var n_ = args.n || '';
    var e_ = args.e || '';
    var cipher_ = args.cipher || '';

    // cast parameters to BigInt
    var n = BigInt(n_);
    var e = BigInt(e_);
    var cipher = BigInt(cipher_);

    console.log("Public Keys");
    console.log("N: " + n);
    console.log("e: " + e);

    console.log("Cripted Message: " + cipher);

    var date = new Date();
    var start = date.getTime();

    var factorization = tdFactors(n);

    date = new Date();
    var end = date.getTime();
    console.log("Time for computing factorization: " + (end - start) + " ms");

    // check if N is a coprime number
    if (factorization.length != 2) {
        console.log("N is not coprime number!");
        return {'Error': 'N is not coprime number!'};
    }

    var p = factorization[0];
    var q = factorization[1];

    console.log("p: " + p + "\n" + "q: " + q);

    var phi = (p - BigInt('1')) * (q - BigInt('1'));

    console.log("phi: " + phi);

    //generate the d decryption exponent
    var d = inverse(e, phi);

    console.log("d: " + d);
    
    //var originalMessage = (cipher ** d)%n;

    //console.log("Original message " + originalMessage);
    
    var body = {
        'p': p.toString(),
        'q': q.toString(),
        'd': d.toString()
    };

    return { 
        statusCode: 200,
        headers: { 'Content-Type': 'application/json' },
        body: {
            "response": body,
            "time": end - start + ' ms'
        }
    }
}

//calculate multiplicative inverse of a%n using the extended euclidean GCD algorithm
function inverse(a, N) {

    var ans = extendedEuclid(a, N);

    if (ans[1] > BigInt('0')) return ans[1];
    else return ans[1] + N;
}

//Calculate d = gcd(a,N) = ax+yN
function extendedEuclid(a, N) {

    var ans = new Array(3);
    var ax, yN;

    if(N == BigInt('0')) {
        ans[0] = a;
        ans[1] = BigInt('1');
        ans[2] = BigInt('0');
        return ans;
    }

    ans = extendedEuclid(N, a%N);
    ax = ans[1];
    yN = ans[2];
    ans[1] = yN;
    var temp = a/N;
    temp = yN * temp;
    ans[2] = ax - temp;
    return ans;
}


// compute factorization
function tdFactors(n) {

    var fs = [];
    if(n < BigInt('2')) throw new IllegalArgumentException('arg n must be greater than one');

    while(n%BigInt('2') == 0) {
        fs.push(BigInt('2'));
        n /= BigInt('2');
    }

    if(n > BigInt('1')) {
        var f = BigInt('3');
        while((f*f) <= n) {
            if(n%f == 0) {
                fs.push(f);
                n /= f;
            }
            else {
                f += BigInt('2');
            }
        }
        fs.push(n);
    }
    return fs;
}

class IllegalArgumentException extends Error {
    constructor(message) {
      super(message);
      this.name = 'IllegalArgumentException';
    }
  }

File: Google Cloud Functions/Java/RSA Decryption/src/main/java/com/moratello/rsa.java
package com.moratello;

import com.google.cloud.functions.HttpFunction;
import com.google.cloud.functions.HttpRequest;
import com.google.cloud.functions.HttpResponse;

import com.google.gson.JsonObject;

import java.math.BigInteger;

import java.util.LinkedList;
import java.util.Map;
import java.util.List;

public class rsa implements HttpFunction {
    @Override
    public void service(HttpRequest request, HttpResponse response)
            throws Exception {

        // response
        JsonObject output = new JsonObject();

        // check if public key n is present in the request
        if(!request.getFirstQueryParameter("n").isPresent()) {
            response.setStatusCode(500);
            response.appendHeader("Content-Type", "application/json");
            output.addProperty("Error", "The parameter n must be defined!");
            response.getWriter().write(output.toString());
            return;
        }

        // check if public key e is present in the request
        if(!request.getFirstQueryParameter("e").isPresent()) {
            response.setStatusCode(500);
            response.appendHeader("Content-Type", "application/json");
            output.addProperty("Error", "The parameter e must be defined!");
            response.getWriter().write(output.toString());
            return;
        }

        // check if the crypted message is present in the request
        if(!request.getFirstQueryParameter("cipher").isPresent()) {
            response.setStatusCode(500);
            response.appendHeader("Content-Type", "application/json");
            output.addProperty("Error", "The cipher must be defined!");
            response.getWriter().write(output.toString());
            return;
        }

        // get parameters
        Map<String, List<String>> args = request.getQueryParameters();

        String n_ = args.get("n").get(0);
        String e_ = args.get("e").get(0);
        String cipher_ = args.get("cipher").get(0);

        BigInteger n = new BigInteger(n_);
        BigInteger e = new BigInteger(e_);
        BigInteger cipher = new BigInteger(cipher_);

        System.out.println("Public Keys");
        System.out.println("N: " + n);
        System.out.println("e: " + e);

        System.out.println("Cripted Message: " + cipher + "\n");

        LinkedList factorization = new LinkedList();

        long start = System.currentTimeMillis();

        factorization = tdFactors(n);

        long end = System.currentTimeMillis();
        System.out.println("Time for computing factorization: " + (end - start) + " ms");

        // check if N is a coprime number
        if (factorization.size() != 2) {
            System.out.println("N is not coprime number!");
            response.setStatusCode(500);
            response.appendHeader("Content-Type", "application/json");
            output.addProperty("Error", "N is not coprime number!");
            response.getWriter().write(output.toString());
            return;
        }

        BigInteger p = (BigInteger) factorization.get(0);
        BigInteger q = (BigInteger) factorization.get(1);

        System.out.println("p: " + p + "\n" + "q: " + q);

        BigInteger phi = (p.subtract(BigInteger.ONE)).multiply(q.subtract(BigInteger.ONE));

        System.out.println("phi: " + phi);

        // generate the d decryption exponent
        BigInteger d = inverse(e, phi);

        System.out.println("d: " + d);

        BigInteger originalMessage = cipher.modPow(d, n);

        System.out.println("Original Message: " + originalMessage);

        response.setStatusCode(200);
        response.appendHeader("Content-Type", "application/json");
        output.addProperty("p", p);
        output.addProperty("q", q);
        output.addProperty("Original Message", originalMessage);
        response.getWriter().write(output.toString());
        return;
    }

    //calculate multiplicative inverse of a%n using the extended euclidean GCD algorithm
    public static BigInteger inverse(BigInteger a, BigInteger N) {

        BigInteger[] ans = extendedEuclid(a, N);

        if (ans[1].compareTo(BigInteger.ZERO) == 1)
            return ans[1];
        else return ans[1].add(N);
    }

    //Calculate d = gcd(a,N) = ax+yN
    public static BigInteger[] extendedEuclid(BigInteger a, BigInteger N) {

        BigInteger[] ans = new BigInteger[3];
        BigInteger ax, yN;

        if (N.equals(BigInteger.ZERO)) {
            ans[0] = a;
            ans[1] = BigInteger.ONE;
            ans[2] = BigInteger.ZERO;
            return ans;
        }

        ans = extendedEuclid(N, a.mod(N));
        ax = ans[1];
        yN = ans[2];
        ans[1] = yN;
        BigInteger temp = a.divide(N);
        temp = yN.multiply(temp);
        ans[2] = ax.subtract(temp);
        return ans;
    }

    // compute factorization
    public static LinkedList tdFactors(BigInteger n) {

        BigInteger two = BigInteger.valueOf(2);
        LinkedList fs = new LinkedList();

        if (n.compareTo(two) < 0) {
            throw new IllegalArgumentException("must be greater than one");
        }

        while (n.mod(two).equals(BigInteger.ZERO)) {
            fs.add(two);
            n = n.divide(two);
        }

        if (n.compareTo(BigInteger.ONE) > 0) {
            BigInteger f = BigInteger.valueOf(3);
            while (f.multiply(f).compareTo(n) <= 0) {
                if (n.mod(f).equals(BigInteger.ZERO)) {
                    fs.add(f);
                    n = n.divide(f);
                } else {
                    f = f.add(two);
                }
            }
            fs.add(n);
        }

        return fs;
    }

}
